File: 3291-find-if-array-can-be-sorted/3291-find-if-array-can-be-sorted.py
class Solution:
    def canSortArray(self, nums: List[int]) -> bool:
        setBits=bin(nums[0]).count('1')
        maxSeg,minSeg=nums[0],nums[0]
        maxPrev=float('-inf')
        for i in range(1,len(nums)):
            if bin(nums[i]).count('1')==setBits:
                maxSeg=max(maxSeg,nums[i])
                minSeg=min(minSeg,nums[i])
            else:
                if minSeg<maxPrev:
                    return False
                maxPrev=maxSeg
                maxSeg,minSeg=nums[i],nums[i]
                setBits=bin(nums[i]).count('1')
        if minSeg<maxPrev:
            return False
        return True

File: 3291-find-if-array-can-be-sorted/3291-find-if-array-can-be-sorted.java
class Solution {
    public boolean canSortArray(int[] nums) {
        HashMap<Integer,Integer> hash=new HashMap<>();
        for(int n:nums){
            hash.put(n,Integer.bitCount(n));
        }
        for(int i=0;i<nums.length;i++){
            for(int j=0;j<nums.length-1-i;j++){
                if(nums[j+1]<nums[j]){
                    if(hash.get(nums[j+1])==hash.get(nums[j])){
                        int temp=nums[j];
                        nums[j]=nums[j+1];
                        nums[j+1]=temp;
                    }
                    else{
                        return false;
                    }
                }
            }
        }
        return true;
    }
}